crates/subspace-archiving/src/archiver.rs

```diff
@@ -388,6 +491,30 @@ impl Archiver {
             let segment_item = match self.buffer.pop_front() {
                 Some(segment_item) => segment_item,
                 None => {
+                    // Segment is not full, so we need to produce incremental mappings.
+                    let new_segment_blocks = match segment.items() {
+                        // If there is a BlockContinuation before this Block, we need to produce
+                        // its mappings as well, because they weren't produced by the previous
+                        // produce_segment() or with_initial_state() call.
+                        [.., SegmentItem::BlockContinuation { .. }, SegmentItem::Block { .. }] => {
+                            &segment.items()[segment.len() - 2..]
+                        }
+                        // Otherwise, just produce mappings for the latest Block.
+                        [.., SegmentItem::Block { .. }] => &segment.items()[segment.len() - 1..],
+                        // We can't tell the difference between an archiver initialized with a
+                        // block continuation, and a spill over from a previous segment. So the
+                        // only way to handle both cases without duplication is to skip a trailing
+                        // continuation, then produce mappings for that continuation when the next
+                        // block is added. (Either here or when the segment is full.)
+                        [.., SegmentItem::BlockContinuation { .. }] => &[],
+                        // If we've just finished the previous segment, the only item in the buffer
+                        // is a SegmentHeader, so there are no new blocks to produce mappings for.
+                        [SegmentItem::ParentSegmentHeader(_)] => &[],
+                        _ => unreachable!("Caller always adds a new block; qed. {}", segment),
```

Comment:
Are all the remaining cases really unreachable?

If we're not absolutely sure, I'm happy to make all these patterns skip the mappings (rather than panicking).